lib/howl/ui/list_widget.moon

```diff
@@ -0,0 +1,287 @@
+  new: (@matcher, opts={}) =>
+    super!
+    @text_widget = TextWidget!
+
+    @opts = moon.copy opts
+    with @opts
+      .filler_text or= '~'
+
+    @_max_height = 10 * @text_widget.row_height * 10
+    @_min_height = 1 * @text_widget.row_height * 1
+
+    @_columns = { { } }
+    @_items = {}
+    @page_start_idx = 1
+    @page_size = 1
+    @selected_idx = nil
+    @column_widths = { 1 }
+    @highlight_matches_for = nil
+
+  to_gobject: => @text_widget\to_gobject!
+
+  @property showing: get: => @text_widget.showing
+
+  @property columns:
+    get: => @_columns
+    set: (val) =>
+      val = val or { {} }
+      @_columns = val
+      @_adjust_height!
+
+  @property offset: get: => @page_start_idx
+
+  @property items: get: => @_items
+
+  @property headers:
+    get: => [column.header for column in *@_columns]
+
+  @property has_header:
+    get: =>
+      for header in *@headers
+        return true if header
+      return false
+
+  @property is_multi_column:
+    get: => return #@_columns > 1
+
+  @property has_status:
+    get: => #@_items == 0 or #@_items > @text_widget.height_rows - (@has_header and 1 or 0)
+
+  @property has_items:
+    get: => #@_items > 0
+
+  @property num_columns:
+    get: => #@_columns
+
+  _write_page: =>
+    @text_widget.buffer.text = ''
+    @page_size = @text_widget.height_rows - (@has_status and 1 or 0) - (@has_header and 1 or 0)
+    if @has_items and @page_size < 1
+      error 'insufficient height - cant display any items'
+
+    items = {}
+    last_idx = @page_start_idx + @page_size - 1
+    for idx = @page_start_idx, math.min(last_idx, #@_items)
+      append items, @_items[idx]
+
+    @text_widget.buffer\append StyledTable items, @columns
+
+    for i = 1, last_idx - #@_items
+      @text_widget.buffer\append @opts.filler_text..'\n', 'comment'
+
+    header_offset = @has_header and 1 or 0
+    for lno = 1, #items
+      line = @text_widget.buffer.lines[lno + header_offset]
+      @_highlight_matches line.text, line.start_pos
+
+    @_write_status!
+
+  _highlight_matches: (text, start_pos) =>
+    if not @highlight_matches_for or @highlight_matches_for.is_empty
+      return
+
+    highlighter = self.highlighter or (text) ->
+      Matcher.explain @highlight_matches_for, text
+
+    positions = highlighter text
+    if positions
+      for hl_pos in *positions
+        p = start_pos + hl_pos - 1
+        @text_widget.buffer\style p, p, 'list_highlight'
+
+  _write_status: =>
+    return unless @has_status
+
+    last_idx = @page_start_idx + @page_size - 1
+    if #@_items < last_idx
+      last_idx = #@_items
+
+    status = '(no items)'
+    if last_idx > 0
+      status = "showing #{@page_start_idx} to #{last_idx} out of #{#@_items}"
+      @text_widget.buffer\append '[..] ', 'comment'
+
+    @text_widget.buffer\append status, 'comment'
+
+  _select: (idx) =>
+    if not @has_items
+      @selected_idx = nil
+      @_highlight nil
+      return
+
+    if idx < 1
```

Comment:
`min`/`max`, or at least an `if`/`elseif` here as the second `if` will never hold true if the first one does.